bulk_query_gen/influxdb/influx_devops_common.go

```diff
@@ -78,40 +77,61 @@ func (d *InfluxDevops) maxCPUUsageHourByMinuteNHosts(qi bulkQuerygen.Query, scal
 
 	hostnameClauses := []string{}
 	for _, s := range hostnames {
-		hostnameClauses = append(hostnameClauses, fmt.Sprintf("hostname = '%s'", s))
+		if d.language == InfluxQL {
+			hostnameClauses = append(hostnameClauses, fmt.Sprintf("hostname = '%s'", s))
+		} else {
+			hostnameClauses = append(hostnameClauses, fmt.Sprintf(`r.hostname == "%s"`, s))
+		}
 	}
 
 	combinedHostnameClause := strings.Join(hostnameClauses, " or ")
 
-	v := url.Values{}
-	v.Set("db", d.DatabaseName)
-	v.Set("q", fmt.Sprintf("SELECT max(usage_user) from cpu where (%s) and time >= '%s' and time < '%s' group by time(1m)", combinedHostnameClause, interval.StartString(), interval.EndString()))
+	var query string
+	if d.language == InfluxQL {
+		query = fmt.Sprintf("SELECT max(usage_user) from cpu where (%s) and time >= '%s' and time < '%s' group by time(1m)", combinedHostnameClause, interval.StartString(), interval.EndString())
+	} else { // Flux
+		query = fmt.Sprintf(`from(db:"%s") ` +
+			`|> range(start:%s, stop:%s) ` +
+			`|> filter(fn:(r) => r._measurement == "cpu" and r._field == "usage_user" and (%s)) ` +
+			`|> keep(columns:["_start", "_stop", "_time", "_value"]) ` +
+			`|> window(period:1m) ` +
+			`|> max() ` +
+			`|> yield()`,
+			d.DatabaseName,
+			interval.StartString(), interval.EndString(),
+			combinedHostnameClause)
+	}
+
+	humanLabel := fmt.Sprintf("InfluxDB (%s) max cpu, rand %4d hosts, rand %s by 1m", d.language.String(), nhosts, timeRange)
 
-	humanLabel := fmt.Sprintf("Influx max cpu, rand %4d hosts, rand %s by 1m", nhosts, timeRange)
 	q := qi.(*bulkQuerygen.HTTPQuery)
-	q.HumanLabel = []byte(humanLabel)
-	q.HumanDescription = []byte(fmt.Sprintf("%s: %s", humanLabel, interval.StartString()))
-	q.Method = []byte("GET")
-	q.Path = []byte(fmt.Sprintf("/query?%s", v.Encode()))
-	q.Body = nil
+	d.getHttpQuery(humanLabel, interval.StartString(), query, q)
 }
 
 // MeanCPUUsageDayByHourAllHosts populates a Query with a query that looks like:
 // SELECT mean(usage_user) from cpu where time >= '$DAY_START' and time < '$DAY_END' group by time(1h),hostname
 func (d *InfluxDevops) MeanCPUUsageDayByHourAllHostsGroupbyHost(qi bulkQuerygen.Query, _ int) {
 	interval := d.AllInterval.RandWindow(24 * time.Hour)
 
-	v := url.Values{}
-	v.Set("db", d.DatabaseName)
-	v.Set("q", fmt.Sprintf("SELECT mean(usage_user) from cpu where time >= '%s' and time < '%s' group by time(1h),hostname", interval.StartString(), interval.EndString()))
+	var query string
+	if d.language == InfluxQL {
+		query = fmt.Sprintf("SELECT mean(usage_user) from cpu where time >= '%s' and time < '%s' group by time(1h),hostname", interval.StartString(), interval.EndString())
+	} else {
+		query = fmt.Sprintf(`from(db:"%s") ` +
+			`|> range(start:%s, stop:%s) ` +
+			`|> filter(fn:(r) => r._measurement == "cpu" and r._field == "usage_user") ` +
+			`|> window(every:1h) ` +
+			`|> mean() ` +
+			`|> group(by:["hostname"]) ` +
```

Comment:
if matching the InfluxQL, group()  should happen before the mean().

Reply:
I puzzled over this for a while last night.  It seems like `window()` reorganizes each input table into several tables, one for each hour, by changing the values of `_start` and `_stop`.  Otherwise, it leaves the group key intact, in this case including `hostname` which is a tag key.  So the tables input to `mean()` are per-hour-per-host, which seems correct.

The call to group() after mean() is just to make the output show rows for the same host as adjacent to one another, which is what InfluxQL does.